includes/background-jobs/class-sensei-scheduler-action-scheduler.php

```diff
@@ -0,0 +1,73 @@
+<?php
+/**
+ * This file contains Sensei_Scheduler_Action_Scheduler class.
+ *
+ * @package sensei
+ */
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly.
+}
+
+/**
+ * Implementation of Sensei's scheduler using Action Scheduler.
+ *
+ * @since 3.0.0
+ */
+class Sensei_Scheduler_Action_Scheduler implements Sensei_Scheduler_Interface {
+	const ACTION_SCHEDULER_GROUP = 'sensei-enrolment';
+
+	/**
+	 * Schedule a job to run.
+	 *
+	 * @param Sensei_Background_Job_Interface $job  Job object.
+	 * @param int|null                        $time Time when the job should run. Defaults to now.
+	 *
+	 * @return mixed
+	 */
+	public function schedule_job( Sensei_Background_Job_Interface $job, $time = null ) {
+		if ( null === $time ) {
+			$time = time();
+		}
+
+		$next_scheduled_action = as_next_scheduled_action( $job->get_name(), [ $job->get_args() ], self::ACTION_SCHEDULER_GROUP );
+
+		if ( ! $next_scheduled_action ) {
+			as_schedule_recurring_action( $time, 0, $job->get_name(), [ $job->get_args() ], self::ACTION_SCHEDULER_GROUP );
+		}
+	}
+
+	/**
+	 * Handle running a job and handling its completion lifecycle event.
+	 *
+	 * @param Sensei_Background_Job_Interface $job                 Job object.
+	 * @param callable|null                   $completion_callback Callback to call when job is complete.
+	 */
+	public function run( Sensei_Background_Job_Interface $job, $completion_callback = null ) {
+		$job->run();
+
+		if ( $job->is_complete() ) {
+			$this->cancel_scheduled_job( $job );
+
+			if ( is_callable( $completion_callback ) ) {
+				call_user_func( $completion_callback );
+			}
+		}
+	}
+
+	/**
+	 * Cancel a scheduled job.
+	 *
+	 * @param Sensei_Background_Job_Interface $job Job to schedule.
+	 */
+	public function cancel_scheduled_job( Sensei_Background_Job_Interface $job ) {
+		as_unschedule_all_actions( $job->get_name(), [ $job->get_args() ], self::ACTION_SCHEDULER_GROUP );
+	}
+
+	/**
+	 * Cancel all jobs.
+	 */
+	public function cancel_all_jobs() {
+		as_unschedule_all_actions( null, null, self::ACTION_SCHEDULER_GROUP );
```

Comment:
We discussed this in Slack. I think we decided that because any orphaned jobs in Cron or Action Scheduler would just gracefully fail, it isn't worth it at this time. I think the main concern is around forgetting to put a job in WP Cron's filter.

One potential option later on is to use the filter as a whitelist to catch us when we forget to add an action.